app/src/main/java/org/greenstand/android/TreeTracker/managers/Locations.kt

```diff
@@ -0,0 +1,332 @@
+    }
+
+    fun hasSufficientAccuracy(): Boolean {
+        return currentLocation?.let {
+            it.hasAccuracy() && it.accuracy < ValueHelper.MIN_ACCURACY_DEFAULT_SETTING
+        } ?: false
+    }
+
+    private fun hasLocationPermissions(): Boolean {
+
+        val fineLocationPermission = ContextCompat.checkSelfPermission(
+            context,
+            android.Manifest.permission.ACCESS_FINE_LOCATION
+        )
+        val coarseLocationPermission = ContextCompat.checkSelfPermission(
+            context,
+            android.Manifest.permission.ACCESS_COARSE_LOCATION
+        )
+        return (fineLocationPermission == PackageManager.PERMISSION_GRANTED ||
+                coarseLocationPermission == PackageManager.PERMISSION_GRANTED)
+    }
+}
+
+enum class Accuracy {
+    GOOD,
+    BAD,
+    NONE
+}
+
+fun Location?.accuracyStatus(): Accuracy {
+    if (this == null || !hasAccuracy()) {
+        return Accuracy.NONE
+    }
+    return if (accuracy < ValueHelper.MIN_ACCURACY_DEFAULT_SETTING) {
+        Accuracy.GOOD
+    } else {
+        Accuracy.BAD
+    }
+}
+
+class LocationDataCapturer(
+    private val userManager: UserManager,
+    private val locationUpdateManager: LocationUpdateManager,
+    private val treeTrackerDAO: TreeTrackerDAO
+) {
+    var convergenceWithinRange: Boolean = false
+
+    private val gson = GsonBuilder().serializeNulls().create()
+    private var locationsDeque: Deque<Location> = LinkedList<Location>()
+    var generatedTreeUuid: UUID? = null
+        private set
+    var convergence: Convergence? = null
+
+    private val locationObserver: Observer<Location?> = Observer {
+        it?.apply {
+
+            MainScope().launch(Dispatchers.IO) {
+                val locationData =
+                    LocationData(
+                        userManager.planterCheckinId,
+                        latitude,
+                        longitude,
+                        accuracy,
+                        generatedTreeUuid?.toString() ?: null,
+                        System.currentTimeMillis()
+                    )
+                val jsonValue = gson.toJson(locationData)
+                Timber.d("Inserting new location data $jsonValue")
+                treeTrackerDAO.insertLocationData(LocationDataEntity(jsonValue))
+            }
+
+            if (isInTreeCaptureMode() && !convergenceWithinRange) {
+
+                var evictedLocation: Location? = if (locationsDeque.size >= CONVERGENCE_DATA_SIZE)
+                    locationsDeque.pollFirst() else null
+                locationsDeque.add(it)
+
+                if (locationsDeque.size >= CONVERGENCE_DATA_SIZE) {
+                    if (convergence == null) {
+                        convergence = Convergence(locationsDeque.toList())
+                        convergence!!.computeConvergence()
+                    } else {
+                        convergence!!.computeSlidingWindowConvergence(evictedLocation!!, it)
```

Comment:
Can we make this something like below to remove all the !!
`
convergence?.let {
   it.computeSlidingWIndow...
} ?: Convergence(...).computeConvergence()
`

Reply:
The `if` condition checks the null value and it is fine to use `!!` in the else (gauranteed to be not null) and also is `terse`.

Reply:
It's fine now, but in the future a NPE can be introduced. If there is a way to avoid !! I think it should be done.

Reply:
The reason why there is a !! is because it's not 100% guaranteed. it's possible the value is changed on another thread in the future. This is why Kotlin suggests using 'let' to check nullability.

Reply:
I get that, updated  to use `convergence?` instead. Thank you.